Added delete() as alias for remove() (ES6 map compatibility)

ulikoehler opened the pull request:
ES6 maps have a `delete()` instead of a `remove()` function, therefore I added `delete()` as an alias for `remove()`. No actually new code has been added, but I added a full unit test (copy of `remove()` unit test) anyway, just to be safe.

Note that there is (as far as I know) no way to achieve 100% compatibility as `Map.prototype.size` is a number whereas `HashMap.prototype.count` is a function. My goal is to get the best compatibility possible without changing anything major in HashMap for now.

flesler commented:
I'm not sure about this, give me a little while to think about it. About the test, rather than copy paste can you instead assign:

```js
HashMap.prototype.delete = HashMap.prototype.remove;
```

And the test:

```js
expect(hashmap.delete).to.equal(hashmap.remove);
```

ulikoehler commented:
Good point, I'll change it. I was looking for a way to just assign the function, but I was not thinking about doing it outside that `HashMap.prototype` assignment.

Regarding the alias, let me tell you a little bit about the usecase.
I use mostly plain old javascript objects (if it's OK to consider keys as strings) or ES6 maps whereever possible. I'm currently doing a project neccessitating to find groups graph nodes that have identical edges.

One part of the algorithm basically uses variable-length 1D `Array`s as keys. ES6 maps will not work properly with array keys unless the arrays are the same object (`===`). This is where `HashMap` comes in. It solves this issue by actually hasing the arrays and provides a semi-drop-in replacement.

As far as I know, the `hashmap` project was created when there was no native ES6 `Map` support in major JS interpreters. I think it's safe to say that `hashmap` is one of the standard solutions for the `There is no real associative array` issue in JS (for usecases where plain objects won't work).

For me, however, it's more of a solution to the `ES6 maps don't work for all usecases` scenario. You might consider that rare, but I'm assuming as the next generation can use ES6 from the start of their programming career, they will get used to the API and expect other things called `Map` to implement pretty much the same API.

So what I'm trying to do here is to remove bottlenecks for people that use ES6 `Map`s already - but also make libraries working on ES6 maps able to use `HashMap` as drop-in replacement. Example: I've implemented a function like [`_.groupBy`](https://lodash.com/docs/4.17.4#groupBy) that creates ES6 maps (for usecases where you need numeric keys, for example). Such functions can only be used on `HashMap` if there is at least basic compatibility between the APIs.

That being said, although it would help to have an API a little bit closer to the ES6 map quasi-standard, I think it's totally OK not to merge this PR. Also, feel free to take your time with this.

flesler commented:
Yes, I think making it as close as possible to the spec is a good thing, a no-brainer except for backwards compatibility.

Let's do this, I think the other way around is better. We'll rename `remove` to `delete` to be closer to the spec, and leave `remove` for now (maybe deprecate it) for back-compat. Can you reverse the code to make `remove` the alias (code, tests, docs). If there's any other method you'd like to migrate, I'm ok with it if we maintain back-compat

ulikoehler commented:
I agree, maintaining backward-compat is absolutely key here. I'm still working on the change you suggested regarding the unit test, but after that I'll move it around so delete is the primary function.

I've checked the [ES6 maps docs](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Map) and I can see three functions which are currently not compatible:
* constructor with 2D list argument: HashMap has a constructor with alternating keys and values. Probably it can be made compatible while maintaining back-compatibility. I'll check if this is feasible.
* `Map.prototype.size`: Would require changes to the internals of `HashMap` to update a size counter on change, but won't affect back-compat negatively.
* [ES6 iterators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/@@iterator). I don't know if there is any way to implement this without requiring support for ES6 in the JS interpreter, but I don't have any experience with implementing iterators so far. I would consider requiring ES6 features etc. as breaking backwards compatibility.

ulikoehler commented:
I've added the alias in the prototype, but regarding the test It didn't work out for me:

```
// expect(hashmap.delete).to.equal(hashmap.remove);
AssertionError: expected [Function: remove] to equal [Function] 
```

I also tried other variants like comparing `hashmap.remove.toString()` but that didn't work either for some reason.

I'll try to work on the other points now, we can fix the unit test later if we can find some way that works.

ulikoehler commented:
Is it OK for you if I submit the other compatibility improvements as separate pull requests?

ulikoehler commented:
I found the issue with the unit tests. Due to chanining which is added ontop of the methods, we can't compare the functions directly.

The only solution I can imagine right now is to add "extra" methods to the prototype that make available the original, non-chained methods. But I believe it's not a good idea to modify the prototype in order to make unit testing a bit easier.

I also noticed that unit tests are missing for the chainability of most of the methods. I will try to add some.

ulikoehler commented:
My mistake, it was just delete() that was missing the chainability unit test. It has been added now.

flesler commented on hashmap.js:
```diff
@@ -166,11 +166,14 @@
 		}
 	};
 
+	// ES6 Map API compatibility
```
The comment is not correct. The API is now compatible, this line of code is rather to keep backwards compatibility with the current codebase

flesler commented on hashmap.js:
```diff
@@ -166,11 +166,14 @@
 	//- Add chaining to all methods that don't return something
 
-	['set','multi','copy','remove','clear','forEach'].forEach(function(method) {
+	['set','multi','copy','delete','remove','clear','forEach'].forEach(function(method) {
```
Rather move `HashMap.prototype.remove = HashMap.prototype.delete;` below this block. We want both functions to be the same

flesler commented on test/test.js:
```diff
@@ -127,8 +128,21 @@ describe('hashmap', function() {
+		});
+	});
+
 	describe('hashmap.remove()', function() {
```
Remember to just this whole method by asserting that it's equal to `delete`. Check comment above on how to make it work

ulikoehler commented:
Thanks for the comments. I've implemented your changes (including the unit test) in https://github.com/flesler/hashmap/pull/34/commits/f877501c03416ff54a36b165f5f8327cadd75c42 .

Also I've changed the terminology in the unit tests from `remove` to `delete` and used `delete` instead of `remove` in every test not specifically requiring the use of `remove` (https://github.com/flesler/hashmap/pull/34/commits/d9283864eb7c6b2ab8a2a3befb8699bc1af7fa38)

flesler commented:
Excellent, I'll check everything once you feel it's all done and merge all together. You don't mind if I squash all your commits into one before merging, right?

ulikoehler commented:
Thanks for reviewing, squashing is totally fine with me! I'm still working on the other PRs, I'll inform you once all of them are ready.

ulikoehler commented:
I think all 3 PRs are ready now. Thanks for the very helpful review and suggestions.

If that makes it easier for you, I can merge all PRs into one to avoid merge conflicts on your side.

flesler commented:
I guess one PR both can be even better. Thank you

ulikoehler commented:
Alright, I'll close these ones and create a new one.
